File: kreuzberg/packages/php/src/Types/Keyword.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types;

/**
 * Extracted keyword with score.
 *
 * @property-read string $text Keyword text
 * @property-read float $score Keyword relevance score (0.0 to 1.0)
 * @property-read string|null $algorithm Algorithm used to extract this keyword ("yake" or "rake")
 * @property-read array<int>|null $positions Character offsets where keyword appears in text
 */
readonly class Keyword
{
    /**
     * @param array<int>|null $positions
     */
    public function __construct(
        public string $text,
        public float $score,
        public ?string $algorithm = null,
        public ?array $positions = null,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        /** @var string $text */
        $text = $data['text'] ?? '';

        $score = 0.0;
        if (isset($data['score'])) {
            $value = $data['score'];
            if (is_numeric($value)) {
                $score = (float) $value;
            }
        }

        /** @var string|null $algorithm */
        $algorithm = $data['algorithm'] ?? null;

        /** @var array<int>|null $positions */
        $positions = $data['positions'] ?? null;

        return new self(
            text: $text,
            score: $score,
            algorithm: $algorithm,
            positions: $positions,
        );
    }
}

File: kreuzberg/packages/elixir/lib/kreuzberg/keyword.ex
defmodule Kreuzberg.Keyword do
  @moduledoc """
  Structure representing an extracted keyword with score and algorithm info.

  ## Fields

    * `:text` - The keyword text
    * `:score` - Relevance score (algorithm-dependent)
    * `:algorithm` - Algorithm used for extraction (e.g., "yake", "rake")
    * `:positions` - Optional list of positions where keyword appears
  """

  @type t :: %__MODULE__{
          text: String.t(),
          score: float(),
          algorithm: String.t(),
          positions: list(non_neg_integer()) | nil
        }

  defstruct [
    :positions,
    text: "",
    score: 0.0,
    algorithm: "yake"
  ]

  @doc """
  Create a Keyword struct from a map.

  ## Examples

      iex> Kreuzberg.Keyword.from_map(%{
      ...>   "text" => "elixir",
      ...>   "score" => 0.95,
      ...>   "algorithm" => "yake"
      ...> })
      %Kreuzberg.Keyword{
        text: "elixir",
        score: 0.95,
        algorithm: "yake",
        positions: nil
      }
  """
  @spec from_map(map()) :: t()
  def from_map(data) when is_map(data) do
    %__MODULE__{
      text: data["text"] || "",
      score: (data["score"] || 0.0) * 1.0,
      algorithm: data["algorithm"] || "yake",
      positions: data["positions"]
    }
  end

  @doc """
  Convert a Keyword struct to a map.

  ## Examples

      iex> kw = %Kreuzberg.Keyword{text: "elixir", score: 0.95, algorithm: "yake"}
      iex> Kreuzberg.Keyword.to_map(kw)
      %{
        "text" => "elixir",
        "score" => 0.95,
        "algorithm" => "yake",
        "positions" => nil
      }
  """
  @spec to_map(t()) :: map()
  def to_map(%__MODULE__{} = kw) do
    %{
      "text" => kw.text,
      "score" => kw.score,
      "algorithm" => kw.algorithm,
      "positions" => kw.positions
    }
  end
end
